src/node.cc

```diff
@@ -1272,6 +1277,141 @@ void SetupNextTick(const FunctionCallbackInfo<Value>& args) {
   args.GetReturnValue().Set(Uint32Array::New(array_buffer, 0, fields_count));
 }
 
+// Data struct that is attached to promises as raw memory.
+struct promise_abort_info {
+#ifdef __POSIX__
+  int fd;
+  pid_t pid;
+  int refcount;
+#endif
+};
+
+void OpenCoreDump(Environment* env, Local<Promise> promise) {
+  if (!abort_on_unhandled_rejection) return;
+#ifdef __POSIX__
+  struct rlimit lim;
+  CHECK_EQ(getrlimit(RLIMIT_CORE, &lim), 0);
+  if (lim.rlim_cur == 0) {
+    // Don't bother creating processes if we cannot create core dumps anyways.
+    return;
+  }
+  // Look up whether there is already a core dump holder process for this
+  // rejection reason, to better capture failure propagation along Promise
+  // chains.
+  Local<NativeWeakMap> reason_map = env->promise_reject_reason_map();
+  if (reason_map.IsEmpty()) {
+    reason_map = NativeWeakMap::New(env->isolate());
+    env->set_promise_reject_reason_map(reason_map);
+  }
+  CHECK_EQ(promise->State(), Promise::kRejected);
+  Local<Value> other = reason_map->Get(promise->Result());
+  if (other->IsPromise()) {
+    Local<Value> uint8array =
+        other.As<Promise>()->GetPrivate(env->context(),
+                                        env->promise_abort_info_symbol())
+                                            .ToLocalChecked();
+    CHECK(uint8array->IsUint8Array());
+    promise_abort_info* info =
+        reinterpret_cast<promise_abort_info*>(Buffer::Data(uint8array));
+    info->refcount++;
+    promise->SetPrivate(env->context(),
+                        env->promise_abort_info_symbol(),
+                        uint8array).FromJust();
+    return;
+  }
+  reason_map->Set(promise->Result(), promise);
+  int pipes[2] = { -1, -1 };
+  pid_t pid;
+
+  if (pipe(pipes) == -1) return;
+  pid = fork();
+  if (pid == -1) {
```

Comment:
Am I reading this correctly that if we fail to fork (or do any of these operations), we just won't get a core file for that, and the user will never know?  I think the documentation should probably reflect that this is a best-effort approach.  It would also be great if there were some way for a user to know that a dump has been skipped because of a failure like this.  I'm afraid that when combined with some of the problems I mentioned earlier (e.g., swap exhaustion, or exhaustion of some other cap -- like number of processes), users won't get a core file, and they won't know that they won't get one until it's too late.

Reply:
In this case, the original process itself will `abort()` once it finds that there is no `promise_abort_info` struct attached to the promise when it would otherwise signal the child process to abort:

https://github.com/addaleax/node/blob/29fcd36f789a2f9913661ba7f6b964362517c9b6/src/node.cc#L1369-L1373

It’s certainly not ideal, but what’s the alternative?

Reply:
I think that aborting the original process if we can't be sure is the reasonable approach.